### PR TITLE
support HTTPS/TLS.

### PR DESCRIPTION
**Pull request by karezachen**
#930 

## HTTP

Demo

```python
from prometheus_client import start_http_server

start_http_server(8000)

while True:
    pass
```

Curl HTTPS Request

```shell
curl -k https://127.0.0.1:8000
```

Response

```shell
curl: (35) LibreSSL/3.3.6: error:1404B42E:SSL routines:ST_CONNECT:tlsv1 alert protocol version
```

### Support HTTPS/TLS

Demo

```python
from prometheus_client import start_http_server

start_http_server(8000, certfile="server.crt", keyfile="server.key")

while True:
    pass
```

Generate server cert and server key.

```shell
openssl genrsa -out server.key 2048
openssl req -new -key server.key -out server.csr
openssl x509 -req -in server.csr -signkey server.key -out server.crt
```

Curl HTTPS Request

```shell
curl -k https://127.0.0.1:8000
```

Response

```shell
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 618.0
python_gc_objects_collected_total{generation="1"} 5.0
python_gc_objects_collected_total{generation="2"} 0.0
# HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 39.0
python_gc_collections_total{generation="1"} 3.0
python_gc_collections_total{generation="2"} 0.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="11",patchlevel="5",version="3.11.5"} 1.0
```

### Enabled mTLS

Demo

```python
from prometheus_client import start_http_server

start_http_server(8000, certfile="server.crt", keyfile="server.key", client_cafile="client.crt", client_auth_required=True)

while True:
    pass
```

Generate client cert and client key.

```shell
openssl genrsa -out client.key 2048
openssl req -new -key client.key -out client.csr
openssl x509 -req -in client.csr -signkey client.key -out client.crt
```

Curl HTTPS Request

```shell
curl -k https://127.0.0.1:8000
```

Response

```shell
curl: (56) LibreSSL SSL_read: LibreSSL/3.3.6: error:1404C45C:SSL routines:ST_OK:reason(1116), errno 0
```

Curl HTTPS Request with client cert and client key

```shell
curl -k --cert client.crt --key client.key https://127.0.0.1:8000
```

Response

```shell
# HELP python_gc_objects_collected_total Objects collected during gc
# TYPE python_gc_objects_collected_total counter
python_gc_objects_collected_total{generation="0"} 618.0
python_gc_objects_collected_total{generation="1"} 5.0
python_gc_objects_collected_total{generation="2"} 0.0
# HELP python_gc_objects_uncollectable_total Uncollectable objects found during GC
# TYPE python_gc_objects_uncollectable_total counter
python_gc_objects_uncollectable_total{generation="0"} 0.0
python_gc_objects_uncollectable_total{generation="1"} 0.0
python_gc_objects_uncollectable_total{generation="2"} 0.0
# HELP python_gc_collections_total Number of times this generation was collected
# TYPE python_gc_collections_total counter
python_gc_collections_total{generation="0"} 39.0
python_gc_collections_total{generation="1"} 3.0
python_gc_collections_total{generation="2"} 0.0
# HELP python_info Python platform information
# TYPE python_info gauge
python_info{implementation="CPython",major="3",minor="11",patchlevel="5",version="3.11.5"} 1.0
```

### CODE REVIEW

**Comment by xiqing-zhang** [comment]
@karezachen Hi, the feature is very nice. We are looking for prometheus python client with TLS support. Our project is pending on the feature. Can you pls set reviewer and ask them to review and merge it?

**Review by xiqing-zhang — APPROVED** [review]
It looks good! Simple and but nice!

**Comment by roidelapluie** [comment]
Thank you. I think this is interesting, however we would need to document it better and make sure that we error if the arguments passed are not correct (e.g. an array of 2). Maybe we should review the API to make it more clear which is which in the array, maybe by using multiple arguments or a custom object? What would be alternate proposals here?

**Comment by Charles1000Chen** [comment]
@karezachen Hi Kareza, thank you for the improvement. 
1. Since Python version 3.10: SSLContext without protocol argument is deprecated, we could use `ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)` instead. 

```
ssl.PROTOCOL_TLS_SERVER
Auto-negotiate the highest protocol version that both the client and server support, and configure the context server-side connections
```

2. About the `context.load_cert_chain(tls[0], tls[1], tls[2] if len(tls) == 3 else None)`, its accepted parameters should be 
```
SSLContext.load_cert_chain(certfile, keyfile=None, password=None)
```

3. To use mTLS to verify client cert, we may want to cal the `load_verify_locations` to set ca cert for the SSLContext object before call `wrap_socket`.
```
SSLContext.load_verify_locations(cafile=None, capath=None, cadata=None)
```

**Comment by karezachen** [comment]
> @karezachen Hi Kareza, thank you for the improvement.
> 
> 1. Since Python version 3.10: SSLContext without protocol argument is deprecated, we could use `ssl.SSLContext(ssl.PROTOCOL_TLS_SERVER)` instead.
> 
> ```
> ssl.PROTOCOL_TLS_SERVER
> Auto-negotiate the highest protocol version that both the client and server support, and configure the context server-side connections
> ```
> 
>  2. About the `context.load_cert_chain(tls[0], tls[1], tls[2] if len(tls) == 3 else None)`, its accepted arguments should be
> 
> ```
> SSLContext.load_cert_chain(certfile, keyfile=None, password=None)
> ```
> 
>  3. To use mTLS to verify client cert, we may want to cal the `load_verify_locations` to set ca cert for the SSLContext object before call `wrap_socket`.
> 
> ```
> SSLContext.load_verify_locations(cafile=None, capath=None, cadata=None)
> ```

**Comment by karezachen** [comment]
@Charles1000Chen @roidelapluie Hi,Thank you for the review.

I have wrapped the `_get_ssl_ctx` method based on the `tls_auth_handler` to support TLS.

If you have any other suggestions, I am more than willing to make the changes.

**Comment by karezachen** [comment]
@csmarchbanks PTAL

File: README.md
Comment (xiqing-zhang) [inline README.md:321]: @karezachen Suggest to add 'cafile' & 'insecure_skip_verify' as optional parameter into the doc.

**Review by xiqing-zhang — CHANGES_REQUESTED** [review]
Doc improvement suggestion. Pls add the optional parameters into the doc. thanks

File: prometheus_client/exposition.py
Comment (andy-maier) [inline prometheus_client/exposition.py:189]: It would be nice if we could in addition (but instead of `cafile`) specify a CA directory via the `capath` parameter of `ssl_cxt.load_verify_locations()`.

File: prometheus_client/exposition.py
Comment (andy-maier) [inline prometheus_client/exposition.py:192]: Putting the new parameters before the previously existing `registry` parameter is an incompatibility for users that pass positional arguments. I think we should stay compatible if possible, i.e. add the new parameters after the previously existing ones.

I think a `*` before the new parameters would also not hurt to require passing them as kwargs.

File: README.md
Comment (andy-maier) [inline README.md:317]: In addition, I think we should describe how the use of mTLS is controlled when enabling HTTPS.

How about:

```
When HTTPS is enabled, if `cafile` is provided, mutual TLS (mTLS) is used and enforced,
i.e. the prometheus client requires that Prometheus presents a client certificate during
the TLS handshake, which the prometheus client validates using the specified CA certificates.
```

**Comment by andy-maier** [comment]
I have started using the PR in its current state for our Z HMC Prometheus exporter (see PR https://github.com/zhmcclient/zhmc-prometheus-exporter/pull/361), and it worked. However, see my comments, above.

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:178]: Update: I tried out the use of `check_hostname = True` and it turns out that this is only for checking the server hostname when you are a client. So in our case, we only need to specify `verify_mode`:

```
    if not insecure_skip_verify:
        ssl_cxt.verify_mode = ssl.CERT_REQUIRED
```

Comment (karezachen) [reply prometheus_client/exposition.py]: I agree with it. DONE

File: README.md
Comment (karezachen) [reply README.md:330]: More detailed documentation would really help others understand and use this feature. Thank you for your review.

File: prometheus_client/exposition.py
Comment (karezachen) [reply prometheus_client/exposition.py:188]: The usage of `str = ''` is used in other methods of this project

e.g.
https://github.com/prometheus/client_python/blob/master/prometheus_client/metrics.py#L116

```python
    def __init__(self: T,
                 name: str,
                 documentation: str,
                 labelnames: Iterable[str] = (),
                 namespace: str = '',
                 subsystem: str = '',
                 unit: str = '',
                 registry: Optional[CollectorRegistry] = REGISTRY,
                 _labelvalues: Optional[Sequence[str]] = None,
                 ) -> None:
```

And this usage of `str = None` fails [mypy](https://github.com/python/mypy) check.

File: README.md
Comment (karezachen) [reply README.md:334]: Great suggestion.

File: README.md
Comment (roidelapluie) [inline README.md:321]: This is incorrect, it is only when a CA file is also present

Comment (andy-maier) [reply README.md:321]: My understanding is that when the `cafile` parameter is not specified (i.e. `None`), the default CA certificate chain is used as described in https://docs.python.org/3/library/ssl.html#ssl.SSLContext.load_default_certs.

The enforcement of mTLS is done by setting `ssl_cxt.verify_mode = ssl.CERT_REQUIRED`, and this is done regardless of `cafile`(or `capath`) being set. The only conditions are that HTTPS was selected and `insecure_skip_verify=False`.

File: README.md
Comment (andy-maier) [inline README.md:325]: 1. Missing back quote before `insecure_skip_verify=True`.

2. As per the current code, `insecure_skip_verify=True` also does not require a client certificate, so I suggest to change the sentence to simply: "can be used to disable mTLS".
validations.

File: README.md
Comment (andy-maier) [inline README.md:328]: Please add some text about the new `capath` parameter, e.g. by adding this sentence after the first sentence: 

```
`capath` can be used to specify a certificate directory containing a CA certificate chain that
is used to validate the client certificate. If neither of them is provided, ...
```

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:189]: The mypy issue should be resolved by specifying:
```
certfile: Optional[str] = None,
keyfile: Optional[str] = None,
```

However, I do understand your comment about other uses of the empty string in the project. Do any of the project maintainers have an opinion on whether to use `None` or empty string to indicate that no file has been provided? @roidelapluie

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:190]: Was added. Thanks!!

File: prometheus_client/exposition.py
Comment (andy-maier) [inline prometheus_client/exposition.py:172]: I think that with the introduction of `capath`, this condition should be changed to:
```
if cafile or capath:
```

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py]: Was done. Thanks!

**Comment by andy-maier** [comment]
@karezachen  Hi Kareza. I have reviewed the latest version of the PR and have opened some new review comments. I was not able to mark some of the old ones as resolved, so I simply added a comment that they were addressed.

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:180]: The actual problem is that the `FileNotFoundError` exception text does not contain the name of the file that is not found:
```
>>> ssl_cxt.load_verify_locations('xx')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
FileNotFoundError: [Errno 2] No such file or directory
```

Same for `PermissionError`.

That makes it hard to find out from the exception message whether certfile, keyfile, or cafile had an issue. (Note: Unfortunately, a missing or unpermitted capath directory gets raised as an issue only later).
 
If you specify your own exception classes, the file name can be part of the exception message.

For example:

```
class PrometheusClientError:
    pass

class PrometheusClientSSLError(PrometheusClientError, ssl.SSLError):
    pass
    
class PrometheusClientFileNotFoundError(PrometheusClientError, FileNotFoundError):
    pass
    
class PrometheusClientPermissionError(PrometheusClientError, PermissionError):
    pass
```

and then:
```
def _get_ssl_ctx(

    . . .

    if cafile is not None or capath is not None:
        try:
            ssl_cxt.load_verify_locations(cafile, capath)
        except IOError as exc:
            exc_type = type(exc)
            msg = str(exc)
            raise exc_type(f"Cannot load CA certificate chain from file {cafile!r} or directory {capath!r}: {msg}")
    else:
        try:
            ssl_cxt.load_default_certs(purpose=ssl.Purpose.CLIENT_AUTH)
        except IOError as exc:
            exc_type = type(exc)
            msg = str(exc)
            raise exc_type(f"Cannot load default CA certificate chain: {msg}")

    . . .

    try:
        ssl_cxt.load_cert_chain(certfile=certfile, keyfile=keyfile)
        except IOError as exc:
            exc_type = type(exc)
            msg = str(exc)
            raise exc_type(f"Cannot load default server certificate file {certfile!r} or its private key file {keyfile!r}: {msg}")    
```

Comment (andy-maier) [reply prometheus_client/exposition.py:180]: I have addressed that in https://github.com/andy-maier/client_python/tree/andy/review-add-tls . Feel free to cherry-pick it from there (just the last commit).

File: README.md
Comment (andy-maier) [reply README.md:339]: I have addressed that in https://github.com/andy-maier/client_python/tree/andy/review-add-tls . Feel free to cherry-pick it from there (just the last commit).

File: README.md
Comment (andy-maier) [reply README.md:342]: I have addressed that in https://github.com/andy-maier/client_python/tree/andy/review-add-tls . Feel free to cherry-pick it from there (just the last commit).

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:172]: I have addressed that in https://github.com/andy-maier/client_python/tree/andy/review-add-tls . Feel free to cherry-pick it from there (just the last commit).

**Comment by karezachen** [comment]
@roidelapluie @xiqing-zhang @xiqing-zhang 

Thank you all for the discussion and review of this PR. What is still missing before it can be merged?

**Comment by andy-maier** [comment]
@karezachen Hello Kareza. Thanks much for integrating the remaining changes. My review comments have all been resolved. From my perspective, the PR could be merged.

@roidelapluie @csmarchbanks Anything left to be done from your perspective?

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py]: Was integrated - resolved.

File: README.md
Comment (andy-maier) [reply README.md]: Was integrated - resolved.

File: README.md
Comment (andy-maier) [reply README.md]: Was integrated - resolved.

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py]: Was integrated - resolved.

**Comment by Charles1000Chen** [comment]
@karezachen @roidelapluie @csmarchbanks  I have tested by zhmc-prometheus-exporter and the mTLS feature works well. Could you help to get the pr merged? Thanks.

**Comment by csmarchbanks** [comment]
👋 I am just coming back from leave and still catching up. I hope to have time to review this soon though if @roidelapluie does not get to it first.

**Comment by GeluConstantin** [comment]
Hi, 

I tired the HTTPS option but I get an error, TypeError: start_wsgi_server() got an unexpected keyword argument 'tls'

it seems that the tls=[certfile, keyfile] is not a valid argument for the function.

Can you advise on the topic?

Thank you,

Constantin

**Comment by karezachen** [comment]
@GeluConstantin 
Hello Gelu,

I did use the `tls` parameter in the [commit](https://github.com/prometheus/client_python/pull/946/commits/f622b22db2a664302a5916114b8347473a6087dd), but subsequent commits have replaced it with other parameters.

Please refer to [README](https://github.com/karezachen/client_python#http) to use `HTTPS` function.

**Comment by andy-maier** [comment]
@GeluConstantin  Hi Constantin. I tried the branch of this PR with HTTPS shortly after its last update and it worked fine. Note that the branch has 7 commits, so you need to use all of them.

**Comment by GeluConstantin** [comment]
> @GeluConstantin Hi Constantin. I tried the branch of this PR with HTTPS shortly after its last update and it worked fine. Note that the branch has 7 commits, so you need to use all of them.

Hi Andy, 

do you know when the PR is going to be integrated into a release? At the moment I use the latest prometheus-client release 0.17.1 and this release has not yet the HTTPS feature.

Thank you for the feebdack,

Constantin

File: README.md
Comment (csmarchbanks) [reply README.md:335]: @roidelapluie I would like to keep this as similar to other exporters as possible, do they require mTLS by default?

**Comment by andy-maier** [comment]
@GeluConstantin No, I don't know. I am not a maintainer of the project.

@roidelapluie or @csmarchbanks Any chance to get this PR reviewed and merged?

**Comment by GeluConstantin** [comment]
I used fastAPI to provide HTTPS, so my problem is solved.

Kind regards,

Gelu

**Comment by csmarchbanks** [comment]
I am out this week, so unlikely until next week. I have the open question about mTLS behavior when I reviewed it, but otherwise this looks ok to me. I just either need to hear back from @roidelapluie or investigate the behavior next week when I am back.

**Comment by Charles1000Chen** [comment]
@csmarchbanks Any result on your investigation on the PR? We do need your help to have this PR merged. Thanks.

File: prometheus_client/exposition.py
Comment (csmarchbanks) [inline prometheus_client/exposition.py:210]: I wonder if it is worth naming these `client_cafile` and `client_capath` to make it very clear that these are for verifying client certificates. Or else update the docstring?

**Review by csmarchbanks — COMMENTED** [review]
Yes, by default mTLS should not be enabled. You can find more detail in https://github.com/prometheus/exporter-toolkit/blob/master/docs/web-configuration.md, but I believe we should not set `ssl_cxt.verify_mode` to `ssl.CERT_REQUIRED` by default, and instead have the user opt in.

To be most like the exporter toolkit we would allow users to pass in any verify mode, but I would also be ok with only the default (`CERT_NONE`), or a flag to enable mTLS and then set `CERT_REQUIRED`.

File: README.md
Comment (andy-maier) [reply README.md:335]: @csmarchbanks @roidelapluie I'm fine with any way to enable mTLS. Let's keep it consistent with the other exporters then.

File: prometheus_client/exposition.py
Comment (andy-maier) [reply prometheus_client/exposition.py:210]: @karezachen I agree that renaming the function parameters as suggested would be a good idea. Would you mind making that change? It would also need to be changed in the README.md file.

**Comment by andy-maier** [comment]
@csmarchbanks @roidelapluie Kareza has updated the PR to implement the suggestion with the renamed client_cafile/path parameters.

It seems the only open comment at this point is the one about the default for mTLS. I suggested a solution in the comment further up. Could you please let us know how you like that solution?

File: README.md
Comment (csmarchbanks) [reply README.md:335]: I am happy with that proposal, and agree that supporting the web config file is outside the scope of this PR, and possibly this package. Let's keep it as simple as possible.

Comment (andy-maier) [reply README.md:335]: @karezachen Hi Kareza. Since Chris has agreed to the proposal, could you please update the PR to implement it:
- Rename the `insecure_skip_verify` parameter to `client_auth_required`.
- This turns around its semantics, so keep the default as `False` but negate any use of it in the code.

**Comment by karezachen** [comment]
@Charles1000Chen @xiqing-zhang @andy-maier @csmarchbanks @roidelapluie 

Thank you very much for your help with this PR, I think our problems have been solved.

Is it possible to merge this PR now? Thanks

File: README.md
Comment (roidelapluie) [reply README.md:335]: Can we just require client auth if a client_cafile is specified?

Comment (karezachen) [reply README.md]: @andy-maier Hi Andy. I have made the following updates:
- Rename the insecure_skip_verify parameter to client_auth_required.
- Update `_get_ssl_ctx` func, default not to set `ssl_ctx.verify_mode=ssl.CERT_REQUIRED`.
- Update the description of default in README.md.
---
And I have updated the Demo in the description of this PR based on the latest code.

Comment (karezachen) [reply README.md]: > Can we just require client auth if a client_cafile is specified?

@andy-maier Andy, I don't quite understand this problem. Can you help me? Thanks.

**Comment by andy-maier** [comment]
@csmarchbanks @roidelapluie Any chance to merge this PR and release a new version? From the perspective of the contributors and reviewers of the PR, all open review comments have been addressed.

**Comment by csmarchbanks** [comment]
This review comment has not been addressed yet, looks like it was hidden by the Github UI though: https://github.com/prometheus/client_python/pull/946#discussion_r1367980465. I do think it is reasonable to remove `client_auth_required` and just toggle the behavior based on if client_cafile is not None.

**Comment by Charles1000Chen** [comment]
> This review comment has not been addressed yet, looks like it was hidden by the Github UI though: [#946 (comment)](https://github.com/prometheus/client_python/pull/946#discussion_r1367980465). I do think it is reasonable to remove `client_auth_required` and just toggle the behavior based on if client_cafile is not None.

@csmarchbanks Hi Chris, the `client_auth_required` has two logics. The one is the `client_auth_required == True` which could be toggled to the behavior based on `cafile is not None or capath is not None` as what your suggested. The other logic is `client_auth_required == True` and`cafile is None and capath is None`, the server will use the certificates from os default certificate store to verify the client. Removing  `client_auth_required` will have the server lose the ability of using os default certificates to verify client. What do you think about this?

**Comment by andy-maier** [comment]
@csmarchbanks As Charles1000Chen points out, removing `client_auth_required` would loose the ability to use the default certificate store. Is that what you want?

**Comment by csmarchbanks** [comment]
Apparently I had a pending comment and didn't click the send button :(

I am happy with the current behavior (keep being able to use the default certificate store). I would like @roidelapluie to ok that as well since he brought up the concern originally though.

**Comment by roidelapluie** [comment]
Let's keep the ability to use the system store. This PR LGTM, thank you all.

**Review by csmarchbanks — APPROVED** [review]
:+1: I think that covers it, thanks as well!

**Comment by andy-maier** [comment]
@karezachen Thanks again, Kareza for the contributions!
@csmarchbanks Thanks for merging the PR! We will now release a beta version of our zhmc prometheus exporter from the master branch.

**Pull request merged**
